test/testGenVectorVc.cxx

```diff
@@ -0,0 +1,304 @@
+
+      for (std::size_t j = 0; j < Vc::double_v::Size; ++j) {
+         ret |= compare(sc.position.x(), vc.position.x()[j]);
+         ret |= compare(sc.position.y(), vc.position.y()[j]);
+         ret |= compare(sc.position.z(), vc.position.z()[j]);
+         ret |= compare(sc.direction.x(), vc.direction.x()[j]);
+         ret |= compare(sc.direction.y(), vc.direction.y()[j]);
+         ret |= compare(sc.direction.z(), vc.direction.z()[j]);
+      }
+   }
+
+   // Now test Transformation3D
+   std::cout << "Transforms :-" << std::endl;
+   for (size_t i = 0; i < nPhotons; ++i) {
+      auto &sc = scalar_data[i];
+      auto &vc = vc_data[i];
+
+      // make 6 random scalar Points
+      Point<double> sp1(p_x(gen), p_y(gen), p_z(gen));
+      Point<double> sp2(p_x(gen), p_y(gen), p_z(gen));
+      Point<double> sp3(p_x(gen), p_y(gen), p_z(gen));
+      Point<double> sp4(p_x(gen), p_y(gen), p_z(gen));
+      Point<double> sp5(p_x(gen), p_y(gen), p_z(gen));
+      Point<double> sp6(p_x(gen), p_y(gen), p_z(gen));
+      // clone to Vc versions
+      Point<Vc::double_v> vp1(sp1.x(), sp1.y(), sp1.z());
+      Point<Vc::double_v> vp2(sp2.x(), sp2.y(), sp2.z());
+      Point<Vc::double_v> vp3(sp3.x(), sp3.y(), sp3.z());
+      Point<Vc::double_v> vp4(sp4.x(), sp4.y(), sp4.z());
+      Point<Vc::double_v> vp5(sp5.x(), sp5.y(), sp5.z());
+      Point<Vc::double_v> vp6(sp6.x(), sp6.y(), sp6.z());
+
+      // Make transformations from points
+      // note warnings about axis not having the same angles expected here...
+      // point is to check scalar and vector versions do the same thing
+      ROOT::Math::Impl::Transform3D<double>       st(sp1, sp2, sp3, sp4, sp5, sp6);
+      ROOT::Math::Impl::Transform3D<Vc::double_v> vt(vp1, vp2, vp3, vp4, vp5, vp6);
+
+      // transform the vectors
+      const auto sv = st * sc.direction;
+      const auto vv = vt * vc.direction;
+      std::cout << "Transformed Direction " << sv << " " << vv << std::endl;
+
+      // invert the transformations
+      st.Invert();
+      vt.Invert();
+
+      // Move the points back
+      const auto sv_i = st * sv;
+      const auto vv_i = vt * vv;
+      std::cout << "Transformed Back Direction " << sc.direction << " " << sv_i << " " << vv_i << std::endl;
+
+      for (std::size_t j = 0; j < Vc::double_v::Size; ++j) {
+         ret |= compare(sv.x(), vv.x()[j]);
+         ret |= compare(sv.y(), vv.y()[j]);
+         ret |= compare(sv.z(), vv.z()[j]);
+         ret |= compare(sc.direction.x(), vv_i.x()[j]);
+         ret |= compare(sc.direction.y(), vv_i.y()[j]);
+         ret |= compare(sc.direction.z(), vv_i.z()[j]);
+      }
+
+      ret |= compare(sc.direction.x(), sv_i.x());
+      ret |= compare(sc.direction.y(), sv_i.y());
+      ret |= compare(sc.direction.z(), sv_i.z());
+   }
+
+   if (ret)
+      std::cerr << "test FAILED !!! " << std::endl;
+   else
+      std::cout << "test OK " << std::endl;
```

Comment:
The point of that test was not to check CPU performance, but to assert that the same results are obtained in both cases. Note in that test the two containers have exactly the same number of entries, so the time to compute each should be more or less the same. The advantage from Vc comes from the fact each entry in  the Vc container is intact a vector type, so has internally `Vc::double_v::Size` data entries in it, that are being processed in parallel. So the effective speed is a factor of this vector size. I guess if you like some sort of speed check could be added.

I actually gave a talk to this to a small LHCb meeting last week. I'll happily upload the sides here. The bottom line is for the nice idealised case I get effectively the perfect speed up from the SIMD address sizes (so 2,4 or 8 depending on the SIMD instruction set used and double or float).

[VectorRayTracing-06032017.pdf](https://github.com/root-project/root/files/841721/VectorRayTracing-06032017.pdf)